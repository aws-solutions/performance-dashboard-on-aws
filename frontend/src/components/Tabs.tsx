/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import React, { useEffect, useState, KeyboardEvent } from "react";
import Tab from "./Tab";
import { ScrollMenu, VisibilityContext } from "react-horizontal-scrolling-menu";
import { LeftArrow, RightArrow } from "./Arrows";

interface Props {
    children: React.ReactNode;
    defaultActive: string;
    showArrows?: boolean;
    ariaLabelledById: string;
}

type scrollVisibilityApiType = React.ContextType<typeof VisibilityContext>;

function Tabs(props: Props) {
    const [scrollMenuObj, setScrollMenuObj] = useState<scrollVisibilityApiType>();
    const [activeTab, setActiveTab] = useState<string>(props.defaultActive);
    const tabsMap = new Map<number, string>();

    useEffect(() => {
        setActiveTab(props.defaultActive);
    }, [props.defaultActive]);

    function getActiveTabIndex(): number {
        for (let [key, value] of Array.from(tabsMap.entries())) {
            if (value === activeTab) {
                return key;
            }
        }
        return 0;
    }

    const onKeyDown = (e: KeyboardEvent<HTMLElement>) => {
        if (e.key === "ArrowRight") {
            const index = getActiveTabIndex();
            if (index < tabsMap.size - 1) {
                setActiveTab(tabsMap.get(index + 1) || props.defaultActive);
            }
        } else if (e.key === "ArrowLeft") {
            const index = getActiveTabIndex();
            if (index > 0) {
                setActiveTab(tabsMap.get(index - 1) || props.defaultActive);
            }
        }
    };

    const activateTabItem = (tab: string, currentTab: HTMLElement) => {
        const rect = currentTab.getBoundingClientRect();
        const wrapper = scrollMenuObj?.scrollContainer?.current;
        if (rect && wrapper) {
            const wrapperRect = wrapper.getBoundingClientRect();
            if (rect.left < wrapperRect.left) {
                scrollMenuObj?.scrollPrev();
            } else {
                const shownWidth = wrapperRect.right - rect.left;
                if (shownWidth < rect.width) {
                    scrollMenuObj?.scrollNext();
                }
            }
        }
        setActiveTab(tab);
    };

    return (
        <div
            className="tabs"
            onKeyDown={onKeyDown}
            role="tablist"
            aria-labelledbyid={props.ariaLabelledById}
        >
            <ScrollMenu
                LeftArrow={props.showArrows && LeftArrow}
                RightArrow={props.showArrows && RightArrow}
                onInit={setScrollMenuObj}
                onWheel={onWheel}
                wrapperClassName="border-base-lighter border-bottom margin-top-1"
            >
                {React.Children.map(props.children, (child: any, index) => {
                    tabsMap.set(index, child.props.id);
                    return (
                        <Tab
                            itemId={child.props.id}
                            activeTab={activeTab}
                            key={child.props.id}
                            label={child.props.label}
                            onClick={activateTabItem}
                            onEnter={activateTabItem}
                        />
                    );
                })}
            </ScrollMenu>

            {React.Children.map(props.children, (child) => {
                const childId = (child as any).props.id;
                if (childId !== activeTab) {
                    return <div id={`${childId}-panel`} role="tabpanel"></div>;
                }
                return (
                    <div
                        id={`${childId}-panel`}
                        className="tab-content"
                        role="tabpanel"
                        tabIndex={0}
                        aria-labelledbyid={`${childId}-tab`}
                    >
                        {(child as any).props.children}
                    </div>
                );
            })}
        </div>
    );
}

function onWheel(apiObj: scrollVisibilityApiType, ev: React.WheelEvent): void {
    const isThouchpad = Math.abs(ev.deltaX) !== 0 || Math.abs(ev.deltaY) < 15;

    if (isThouchpad) {
        ev.stopPropagation();
        return;
    }
    if (ev.deltaY < 0) {
        apiObj.scrollNext();
    } else if (ev.deltaY > 0) {
        apiObj.scrollPrev();
    }
}

export default Tabs;
