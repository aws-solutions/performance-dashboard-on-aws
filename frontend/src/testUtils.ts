/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

export type IntersectionObserverCB = (arg1: IntersectionObserverEntry[]) => void;

export class MockedObserver {
    cb: IntersectionObserverCB;
    options: IntersectionObserverInit;
    elements: HTMLElement[];

    constructor(cb: IntersectionObserverCB, options: IntersectionObserverInit) {
        this.cb = cb;
        this.options = options;
        this.elements = [];
    }

    unobserve(elem: HTMLElement): void {
        this.elements = this.elements.filter((en) => en !== elem);
    }

    observe(elem: HTMLElement): void {
        this.elements = [...Array.from(new Set(this.elements.concat(elem)))];
    }

    disconnect(): void {
        this.elements = [];
    }

    fire(arr: IntersectionObserverEntry[]): void {
        this.cb(arr);
    }
}

export function traceMethodCalls(obj: object | Function, calls: any = {}) {
    const handler: ProxyHandler<object | Function> = {
        get(target, propKey, receiver) {
            const targetValue = Reflect.get(target, propKey, receiver);
            if (typeof targetValue === "function") {
                return function (...args: any[]) {
                    calls[propKey] = (calls[propKey] || []).concat(args);
                    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                    // @ts-ignore
                    return targetValue.apply(this, args);
                };
            } else {
                return targetValue;
            }
        },
    };
    return new Proxy(obj, handler);
}

export class MockedResizeObserver {
    observe() {
        // do nothing
    }
    unobserve() {
        // do nothing
    }
    disconnect() {
        // do nothing
    }
}
