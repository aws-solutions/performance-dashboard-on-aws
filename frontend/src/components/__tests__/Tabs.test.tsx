/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import { render, within } from "@testing-library/react";
import Tabs from "../Tabs";
import { MemoryRouter } from "react-router-dom";
import { MockedObserver, traceMethodCalls, IntersectionObserverCB } from "../../testUtils";

describe("Tabs tests", () => {
    let observer: any;
    let mockedObserverCalls: { [k: string]: any } = {};
    beforeEach(() => {
        Object.defineProperty(window, "IntersectionObserver", {
            writable: true,
            value: jest.fn().mockImplementation(function TrackMock(
                cb: IntersectionObserverCB,
                // eslint-disable-next-line no-undef
                options: IntersectionObserverInit,
            ) {
                observer = traceMethodCalls(new MockedObserver(cb, options), mockedObserverCalls);

                return observer;
            }),
        });
    });
    afterEach(() => {
        observer = null;
        mockedObserverCalls = {};
    });

    test("renders the Tabs component", async () => {
        const wrapper = render(
            <Tabs defaultActive="tab1" container="tab1test">
                <div id="tab1" label="Tab 1">
                    Tab 1
                </div>
                <div id="tab2" label="Tab 2">
                    Tab 2
                </div>
            </Tabs>,
        );
        expect(wrapper.container).toMatchSnapshot();
    });

    test("renders the Tabs component with default tab 2 selected", async () => {
        const wrapper = render(
            <Tabs defaultActive="tab2" container="tab2test">
                <div id="tab1" label="Tab 1">
                    Tab 1
                </div>
                <div id="tab2" label="Tab 2">
                    Tab 2
                </div>
            </Tabs>,
        );
        expect(wrapper.container).toMatchSnapshot();
    });

    test("renders the tabs", async () => {
        const { getAllByRole } = render(
            <Tabs defaultActive="tab2">
                <div id="tab1" label="Tab 1">
                    Tab 1
                </div>
                <div id="tab2" label="Tab 2">
                    Tab 2
                </div>
            </Tabs>,
            {
                wrapper: MemoryRouter,
            },
        );

        const listItems = getAllByRole("tab");
        expect(listItems).toHaveLength(2);

        listItems.forEach((item, index) => {
            const { getByText } = within(item);
            expect(getByText(`Tab ${index + 1}`)).toBeInTheDocument();
        });
    });
});
