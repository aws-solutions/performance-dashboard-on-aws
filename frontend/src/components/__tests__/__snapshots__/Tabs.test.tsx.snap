// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs tests renders the Tabs component 1`] = `
<div>
  <div
    class="tabs"
    role="tablist"
  >
    <div
      class="react-horizontal-scrolling-menu--wrapper border-base-lighter border-bottom margin-top-1"
    >
      <div
        class="react-horizontal-scrolling-menu--scroll-container "
      >
        <div
          class="react-horizontal-scrolling-menu--item "
          data-index="0"
          data-key="tab1"
        >
          <div
            aria-controls="tab1-panel"
            aria-selected="true"
            class="tab display-inline-block padding-x-2 padding-y-105 text-bold font-sans-md"
            id="tab1-tab"
            role="tab"
            tabindex="0"
          >
            Tab 1
          </div>
        </div>
        <div
          class="react-horizontal-scrolling-menu--separator "
          data-index="0.1"
          data-key="tab1-separator"
        />
        <div
          class="react-horizontal-scrolling-menu--item "
          data-index="1"
          data-key="tab2"
        >
          <div
            aria-controls="tab2-panel"
            aria-selected="false"
            class="tab display-inline-block padding-x-2 padding-y-105 text-bold font-sans-md"
            id="tab2-tab"
            role="tab"
            tabindex="-1"
          >
            Tab 2
          </div>
        </div>
      </div>
    </div>
    <div
      aria-labelledbyid="tab1-tab"
      class="tab-content"
      id="tab1-panel"
      role="tabpanel"
      tabindex="0"
    >
      Tab 1
    </div>
    <div
      id="tab2-panel"
      role="tabpanel"
    />
  </div>
</div>
`;

exports[`Tabs tests renders the Tabs component with default tab 2 selected 1`] = `
<div>
  <div
    class="tabs"
    role="tablist"
  >
    <div
      class="react-horizontal-scrolling-menu--wrapper border-base-lighter border-bottom margin-top-1"
    >
      <div
        class="react-horizontal-scrolling-menu--scroll-container "
      >
        <div
          class="react-horizontal-scrolling-menu--item "
          data-index="0"
          data-key="tab1"
        >
          <div
            aria-controls="tab1-panel"
            aria-selected="false"
            class="tab display-inline-block padding-x-2 padding-y-105 text-bold font-sans-md"
            id="tab1-tab"
            role="tab"
            tabindex="-1"
          >
            Tab 1
          </div>
        </div>
        <div
          class="react-horizontal-scrolling-menu--separator "
          data-index="0.1"
          data-key="tab1-separator"
        />
        <div
          class="react-horizontal-scrolling-menu--item "
          data-index="1"
          data-key="tab2"
        >
          <div
            aria-controls="tab2-panel"
            aria-selected="true"
            class="tab display-inline-block padding-x-2 padding-y-105 text-bold font-sans-md"
            id="tab2-tab"
            role="tab"
            tabindex="0"
          >
            Tab 2
          </div>
        </div>
      </div>
    </div>
    <div
      id="tab1-panel"
      role="tabpanel"
    />
    <div
      aria-labelledbyid="tab2-tab"
      class="tab-content"
      id="tab2-panel"
      role="tabpanel"
      tabindex="0"
    >
      Tab 2
    </div>
  </div>
</div>
`;
